third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,876 @@
+                mapped_sym,
+                if sym_is_unsafe {
+                    SimpleValue::NonConstant
+                } else {
+                    SimpleValue::LocalVar(mapped_sym)
+                },
+            );
+        }
+    }
+
+    fn rewrite_exit_scope(&mut self, _id: NodeId) {
+        self.visiting_binding.exit_scope();
+        self.values.exit_scope();
+        self.remapped_symbol.exit_scope();
+    }
+
+    fn rewrite_local_var(&mut self, id: NodeId, sym: Symbol) -> Option<Exp> {
+        self.rewrite_to_recorded_value(id, sym)
+    }
+
+    fn rewrite_assign(&mut self, id: NodeId, lhs: &Pattern, rhs: &Exp) -> Option<Exp> {
+        let mut remove_vars = BTreeSet::new();
+        for (var, opt_exp) in lhs.vars_and_exprs(rhs) {
+            if let Some(exp) = opt_exp {
+                if let ExpData::LocalVar(_, sym) = exp.as_ref() {
+                    if var == *sym {
+                        remove_vars.insert(var);
+                    }
+                }
+            }
+        }
+        if !remove_vars.is_empty() {
+            let new_lhs = lhs.clone().remove_vars(&remove_vars);
+            Some(ExpData::Assign(id, new_lhs, rhs.clone()).into_exp())
+        } else {
+            None
+        }
+    }
+
+    fn rewrite_temporary(&mut self, _id: NodeId, _idx: TempIndex) -> Option<Exp> {
+        // Note that at this phase of compilation, any temporary must refer to
+        // a parameter.  This means that the value is comes from outside and is
+        // unknown, so no rewrite will be useful.
+        None
+    }
+
+    fn rewrite_call(&mut self, id: NodeId, oper: &Operation, args: &[Exp]) -> Option<Exp> {
+        self.constant_folder
+            .rewrite_call(id, oper, args)
+            .map(|exp| self.expand_tuple(exp))
+            .or({
+                // Not completely a constant.
+                // TODO(BUGBUG): consider matching some expressions,
+                // e.g., ((x + c1) + c2) -> (x + (c1 + c2))
+                None
+            })
+    }
+
+    fn rewrite_if_else(&mut self, _id: NodeId, cond: &Exp, then: &Exp, else_: &Exp) -> Option<Exp> {
+        if DISABLE_IF_SIMPLIFICATION {
+            None
+        } else {
+            match cond.as_ref() {
+                ExpData::Value(_, Value::Bool(true)) => Some(then.clone()),
+                ExpData::Value(_, Value::Bool(false)) => Some(else_.clone()),
+                _ => None,
+            }
+        }
+    }
+
+    fn rewrite_sequence(&mut self, id: NodeId, seq: &[Exp]) -> Option<Exp> {
+        if self.eliminate_code {
+            // Check which elements are side-effect-free
+            let mut siter = seq.iter();
+            let _ignore = siter.next_back(); // ignore last element, we have to keep it
+            let mut side_effect_free_elts: Vec<_> = siter
+                .map(|exp| exp.as_ref().is_side_effect_free())
+                .collect();
+            if side_effect_free_elts.iter().all(|elt| *elt) {
```

Comment:
I wonder whether you can simplify the code in this function by just constructing the list of side-effect elements in the sequence. If that length is smaller than the original you rewrite. It appears in the current code you iterate over the sequence more than once, so this should also be faster.